src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs

```diff
@@ -256,6 +256,299 @@ public static JsonDocument Parse(string json, JsonReaderOptions readerOptions =
+        {
+            return TryReadFrom(ref reader, out document, shouldThrow: false);
+        }
+
+        /// <summary>
+        ///   Parses one JSON value (including objects or arrays) from the provided reader.
+        /// </summary>
+        /// <param name="reader">The reader to read.</param>
+        /// <returns>
+        ///   A JsonDocument representing the value (and nested values) read from the reader.
+        /// </returns>
+        /// <remarks>
+        ///   <para>
+        ///     Upon completion of this method <paramref name="reader"/> will positioned at the
+        ///     final token in the JSON value.  If an exception is thrown the reader is reset to
+        ///     the state it was in when the method was called.
+        ///   </para>
+        ///
+        ///   <para>
+        ///     This method makes a copy of the data the reader acted on, there is no caller
+        ///     requirement to maintain data integrity beyond the return of this method.
+        ///   </para>
+        /// </remarks>
+        /// <exception cref="ArgumentException">
+        ///   <paramref name="reader"/> is using unsupported options.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        ///   The current <paramref name="reader"/> token does not start or represent a value.
+        /// </exception>
+        /// <exception cref="JsonReaderException">
+        ///   A value could not be read from <paramref name="reader"/>.
+        /// </exception>
+        public static JsonDocument ReadFrom(ref Utf8JsonReader reader)
+        {
+            bool ret = TryReadFrom(ref reader, out JsonDocument document, shouldThrow: true);
+            Debug.Assert(ret, "TryReadFrom returned false with shouldThrow: true.");
+            return document;
+        }
+
+        private static bool TryReadFrom(ref Utf8JsonReader reader, out JsonDocument document, bool shouldThrow)
+        {
+            JsonReaderState state = reader.CurrentState;
+            CheckSupportedOptions(state.Options, nameof(reader));
+
+            // Value copy to overwrite the ref on an exception and undo the destructive reads.
+            Utf8JsonReader restore = reader;
+
+            // Only used for StartArray or StartObject,
+            // the beginning of the token is one byte earlier.
+            long startingOffset = state.BytesConsumed;
+
+            ReadOnlySpan<byte> valueSpan = default;
+            ReadOnlySequence<byte> valueSequence = default;
+
+            try
+            {
+                switch (reader.TokenType)
+                {
+                    // A new reader was created and has never been read.
+                    // Or a reader has terminated and we're about to throw.
+                    case JsonTokenType.None:
+                    // Using a reader loop the caller has identified a property they wish to
+                    // hydrate into a JsonDocument. Move to the value first.
```

Comment:
It's not clear to me from these comments why the handling for None is the same as the handling for PropertyName…?

Reply:
None mainly means "the reader was created and no one called Read() yet", so it's "go to the next (first) token and ask it what it is"